File: python/data_structures/stack.py
"""
Stack Implementation
Time Complexity: O(1) for all operations
Space Complexity: O(n)
"""

class Stack:
    """Stack data structure implementation using list"""
    
    def __init__(self):
        """Initialize empty stack"""
        self.data = []
    
    def push(self, value):
        """
        Push element to stack
        
        Args:
            value: Element to push
        """
        self.data.append(value)
    
    def pop(self):
        """
        Pop element from stack
        
        Returns:
            Top element, or None if stack is empty
        """
        if self.is_empty():
            print("Stack underflow!")
            return None
        return self.data.pop()
    
    def top(self):
        """
        Get top element without removing it
        
        Returns:
            Top element, or None if stack is empty
        """
        if self.is_empty():
            print("Stack is empty!")
            return None
        return self.data[-1]
    
    def is_empty(self):
        """
        Check if stack is empty
        
        Returns:
            True if stack is empty, False otherwise
        """
        return len(self.data) == 0
    
    def size(self):
        """
        Get stack size
        
        Returns:
            Number of elements in stack
        """
        return len(self.data)
    
    def display(self):
        """Display stack contents"""
        if self.is_empty():
            print("Stack is empty")
            return
        print("Stack:", " ".join(map(str, reversed(self.data))))

if __name__ == "__main__":
    s = Stack()
    
    s.push(10)
    s.push(20)
    s.push(30)
    
    print("After pushing 10, 20, 30:")
    s.display()
    
    print(f"Top element: {s.top()}")
    print(f"Stack size: {s.size()}")
    
    print(f"Popped element: {s.pop()}")
    s.display()

File: cpp/data_structures/stack.cpp
#include <iostream>
#include <vector>
using namespace std;

// Stack Implementation using Vector
// Time Complexity: O(1) for all operations
// Space Complexity: O(n)

class Stack {
private:
    vector<int> data;
    
public:
    // Push element to stack
    void push(int value) {
        data.push_back(value);
    }
    
    // Pop element from stack
    int pop() {
        if (isEmpty()) {
            cout << "Stack underflow!" << endl;
            return -1;
        }
        int value = data.back();
        data.pop_back();
        return value;
    }
    
    // Get top element
    int top() {
        if (isEmpty()) {
            cout << "Stack is empty!" << endl;
            return -1;
        }
        return data.back();
    }
    
    // Check if stack is empty
    bool isEmpty() {
        return data.empty();
    }
    
    // Get stack size
    int size() {
        return data.size();
    }
    
    // Display stack
    void display() {
        if (isEmpty()) {
            cout << "Stack is empty" << endl;
            return;
        }
        cout << "Stack: ";
        for (int i = data.size() - 1; i >= 0; i--) {
            cout << data[i] << " ";
        }
        cout << endl;
    }
};

int main() {
    Stack s;
    
    s.push(10);
    s.push(20);
    s.push(30);
    
    cout << "After pushing 10, 20, 30:" << endl;
    s.display();
    
    cout << "Top element: " << s.top() << endl;
    cout << "Stack size: " << s.size() << endl;
    
    cout << "Popped element: " << s.pop() << endl;
    s.display();
    
    return 0;
}
